openapi.yaml

```diff
@@ -0,0 +1,712 @@
+        '429':
+          description: Too many requests
+        '500':
+          description: Internal server error _(service cannot be deleted)_
+      security:
+        - subscription_key: []
+  /services/{serviceId}/logo:
+    put:
+      tags:
+        - services
+      summary: Upload service logo
+      description: |
+        Upload service logo by by `serviceId`. <br><br>_Note: This API operation needs the __Manage API key__ owned by the service user, otherwise `403 forbidden` will be returned to the caller._
+      operationId: updateServiceLogo
+      parameters:
+        - name: serviceId
+          in: path
+          description: ID of the service
+          required: true
+          schema:
+            type: string
+      requestBody:
+        description: Service logo payload _(base64 string representation)_
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Logo'
+        required: true
+      responses:
+        '200':
+          description: Successful operation _(service logo updated)_
+        '400':
+          description: Invalid payload
+        '401':
+          description: Unauthorized
+        '403':
+          description: Forbidden
+        '404':
+          description: Not found
+        '429':
+          description: Too many requests
+        '500':
+          description: Internal server error _(service logo cannot be updated)_
+      security:
+        - subscription_key: []
+  /services/{serviceId}/key:
+    put:
+      tags:
+        - services-authorization
+      summary: Regenerate service key
+      description: |
+        Regenerate service key by `serviceId`. <br><br>_Note: This API operation needs the __Manage API key__ owned by the service user, otherwise `403 forbidden` will be returned to the caller._
+      operationId: regenerateServiceKey
+      parameters:
+        - name: serviceId
+          in: path
+          description: ID of the service
+          required: true
+          schema:
+            type: string
+      requestBody:
+        description: Service key type
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/SubscriptionKeyTypePayload'
+        required: true
+      responses:
+        '200':
+          description: Successful operation _(service key regenerated)_
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SubscriptionKeys'
+        '400':
+          description: Invalid payload
+        '401':
+          description: Unauthorized
+        '403':
+          description: Forbidden
+        '404':
+          description: Not found
+        '429':
+          description: Too many requests
+        '500':
+          description: Internal server error _(service key cannot be regenerated)_
+      security:
+        - subscription_key: []
+  /services/{serviceId}/actions/review:
+    patch: 
```

Comment:
I am not very "happy" with the design of "actions" because they are not RESTful API.

Reply:
We think that Option 1. is what we currently have.
We agree that "services-actions" could be improved. We suggest:

__services-actions:__
__from__ PATCH /services/{serviceId}/actions/review __to__ POST /services/{serviceId}/review _(Submit a review)_
__from__ PATCH /services/{serviceId}/actions/explain __to__ PATCH/PUT /services/{serviceId}/review _(Explain a review)_

Reply:
I add one more "trade-off" point:
4. we consider the service status a "sub-resources", so we can update it by:
  - `PUT /services/{serviceId}/status/{status}` where `status ∈ {Submitted, Rejected, ecc.}`

I mean, if this isn't the "external" but a kind of "internal" OpenAPI definition, I prefer to design it as point 4, because is more clear to the "domain team". On the other hand, for the "external/proxy" OpenAPI definition, i think the  2nd point is more "understanding"

Reply:
ac63b0f